src/debugpy/server/adapters.py

```diff
@@ -346,6 +350,40 @@ def next_request(self, request: Request):
         self._tracer.step_over(thread)
         return {}
 
+    def gotoTargets_request(self, request: Request) -> dict:
+        path = request("source", dict)["path"]
+        line = request("line", int)
+        target_id = self._goto_targets_map.obtain_key((path, line))
+        target = {"id": target_id, "label": "%s:%s" % (path, line), "line": line}
+
+        return {"targets": [target]}
+    
+
+    def goto_request(self, request: Request) -> dict:
+        thread_id = request("threadId", int)
+        thread = Thread.get(thread_id)
+        if thread is None:
+            return request.isnt_valid(f'Unknown thread with "threadId":{thread_id}')
+        target_id = request("targetId", int)
+        if target_id is None:
+            return request.isnt_valid('Unknown targetId for goto')
+        try:
+            path, line = self._goto_targets_map.obtain_value(target_id)
+        except KeyError:
+            return request.isnt_valid('Invalid targetId for goto')
+
+        # Make sure the thread is in the same source file
+        current_source = inspect.getsourcefile(thread.current_frame.f_code)
+        if current_source.casefold() != path.casefold():
```

Comment:
`casefold()` won't do the right thing for case-sensitive filesystems on Unix.

For breakpoints, I've used `pathlib.Path.resolve()` to normalize paths while also resolving symlinks etc - this is done in class `Source`, and should probably be reused here. In general I think since DAP always uses "source" whenever paths appear, it's best to handle paths throughout the debugger by converting them to `Source` as soon as possible, and then working with those objects - that way it will all be consistent.